controller/AuthoringTool.php

```diff
@@ -84,56 +85,71 @@ protected function getValidatedLtiMessagePayload(): LtiMessagePayloadInterface
      */
     public function launch(): void
     {
-        $ltiMessage = $this->getLtiMessageOrRedirectToLogin();
+        try {
+            $ltiMessage = $this->getValidatedLtiMessagePayload();
 
-        $user = $this->getServiceLocator()
-            ->getContainer()
-            ->get(tao_models_classes_UserService::class)
-            ->addUser(
+            $user = $this->getUserService()->addUser(
                 $ltiMessage->getUserIdentity()->getIdentifier(),
                 helpers_Random::generateString(UserService::PASSWORD_LENGTH),
                 new core_kernel_classes_Resource(
                     $this->getAuthoringRoleService()->getValidRole($ltiMessage->getRoles())
                 )
             );
 
-        $this->getServiceLocator()
-            ->getContainer()
-            ->get(LtiService::class)
-            ->startLti1p3Session($ltiMessage, $user);
+            $this->getLtiService()->startLti1p3Session($ltiMessage, $user);
+        } catch (LtiException $exception) {
+            $this->handleLtiException($exception);
+        }
 
-        $this->forward('run', null, null, $_GET);
+        // Using a 301 Moved Permanently redirect prevents getting the request recorded in the browser's
+        // history, avoiding going back to the token validation when the user clicks the "Back" button
+        $this->redirect(_url('entry', 'Main', 'tao', $_GET), 301);
     }
 
     /**
-     * @throws ContainerExceptionInterface
      * @throws InterruptedActionException
      * @throws LtiException
-     * @throws NotFoundExceptionInterface
      */
-    private function getLtiMessageOrRedirectToLogin(): LtiMessagePayloadInterface
+    private function handleLtiException(LtiException $exception): void
     {
-        try {
-            $message = $this->getValidatedLtiMessagePayload();
-        } catch (LtiException $exception) {
-            if ($exception->getMessage() !== self::LTI_NO_MATCHING_REGISTRATION_FOUND_MESSAGE) {
-                throw $exception;
-            }
-
+        if ($exception->getMessage() === self::LTI_NO_MATCHING_REGISTRATION_FOUND_MESSAGE) {
             $this->getLogger()->warning(
                 sprintf(
                     'Missing registration for current audience. Redirecting to the login page. Exception: %s',
                     $exception
                 )
             );
-            $this->redirect(_url('login', 'Main', 'tao'));
+
+            $this->redirect(_url('login', 'Main', 'tao'));  // throws
         }
 
-        return $message;
+        throw $exception;
```

Comment:
I am not really sure it is a good thing to perform a throw after a redirection. It is a bit confusing as both represent a end of a flow. 

Would it make sense to only throw the exception in case there is no redirection, and in that case, only log it?

Reply:
Previously we had this code:

```php
            if ($exception->getMessage() !== self::LTI_NO_MATCHING_REGISTRATION_FOUND_MESSAGE) {
                throw $exception;
            }
```

That was meant to _not hide_ an unexpected exception.

Throwing here serves to keep that old behavior: If there is an `LtiException` but it is not due to using an invalid registration (i.e. the comparison with `LTI_NO_MATCHING_REGISTRATION_FOUND_MESSAGE`), we rethrow the exception, and let an upper-level exception handling block to catch it in order to show the "Launch error" page.

Regarding the redirection, the call to `$this->redirect(_url('login', 'Main', 'tao'));` throws an exception (I was indicating that with a `// throws` comment, since otherwise it may not be immediately clear). That exception causes a `catch` block in an upper level to stop handling the request and send the redirect (i.e. the call to `$this->redirect()` will never return to this controller).

Reply:
`redirect()` is implemented here: https://github.com/oat-sa/tao-core/blob/master/models/classes/http/HttpFlowTrait.php#L59

Reply:
Can we put a return after the redirect? Does not make sense to redirect and after thrown an exception.

Reply:
Inverting the condition here: 08fae81

Reply:
Thanks so much! :)